flink-java/src/main/java/org/apache/flink/api/java/DataSet.java

```diff
@@ -1223,6 +1230,51 @@ public long count() throws Exception {
 		final TypeInformation<K> keyType = TypeExtractor.getKeySelectorTypes(keyExtractor, getType());
 		return new PartitionOperator<T>(this, PartitionMethod.HASH, new Keys.SelectorFunctionKeys<T, K>(clean(keyExtractor), this.getType(), keyType), Utils.getCallLocationName());
 	}
+
+	/**
+	 * Range-partitions a DataSet using the specified KeySelector.
+	 * <p>
+	 * <b>Important:</b>This operation shuffles the whole DataSet over the network and can take significant amount of time.
+	 *
+	 * @param keySelector The KeySelector with which the DataSet is range-partitioned.
+	 * @return The partitioned DataSet.
+	 *
+	 * @see KeySelector
+	 */
+	public <K extends Comparable<K>> DataSet<T> partitionByRange(KeySelector<T, K> keySelector) {
+		final TypeInformation<K> keyType = TypeExtractor.getKeySelectorTypes(keySelector, getType());
+		String callLocation = Utils.getCallLocationName();
+
+		// Extract key from input element by keySelector.
+		KeyExtractorMapper<T, K> keyExtractorMapper = new KeyExtractorMapper<T, K>(keySelector);
```

Comment:
In `connectJobVertices()`, the source and target `JobVertex` has already been created, so just rewrite the `OptimizedPlan` does not work here, unless we generate related `JobVetex` manually which is what i try to avoid, we need to rewrite the `OptimizedPlan` before the creating of related `JobVertex`.